src/oscar/apps/basket/middleware.py

```diff
@@ -19,78 +20,86 @@ def __init__(self, get_response):
-            Return the basket after applying offers.
-            """
-            basket = self.get_basket(request)
-            basket.strategy = request.strategy
-            self.apply_offers_to_basket(request, basket)
-
-            return basket
-
-        def load_basket_hash():
-            """
-            Load the basket and return the basket hash
-
-            Note that we don't apply offers or check that every line has a
-            stockrecord here.
-            """
-            basket = self.get_basket(request)
-            if basket.id:
-                return self.get_basket_hash(basket.id)
-
-        # Use Django's SimpleLazyObject to only perform the loading work
-        # when the attribute is accessed.
-        request.basket = SimpleLazyObject(load_full_basket)
-        request.basket_hash = SimpleLazyObject(load_basket_hash)
+        if self.is_basket_enabled(request):
+            # Keep track of cookies that need to be deleted (which can only be done
+            # when we're processing the response instance).
+            request.cookies_to_delete = []
+
+            # Load stock/price strategy and assign to request (it will later be
+            # assigned to the basket too).
+            strategy = selector.strategy(request=request, user=request.user)
+            request.strategy = strategy
+
+            # We lazily load the basket so use a private variable to hold the
+            # cached instance.
+            request._basket_cache = None
+
+            def load_full_basket():
+                """
+                Return the basket after applying offers.
+                """
+                basket = self.get_basket(request)
+                basket.strategy = request.strategy
+                self.apply_offers_to_basket(request, basket)
+
+                return basket
+
+            def load_basket_hash():
+                """
+                Load the basket and return the basket hash
+
+                Note that we don't apply offers or check that every line has a
+                stockrecord here.
+                """
+                basket = self.get_basket(request)
+                if basket.id:
+                    return self.get_basket_hash(basket.id)
+
+            # Use Django's SimpleLazyObject to only perform the loading work
+            # when the attribute is accessed.
+            request.basket = SimpleLazyObject(load_full_basket)
+            request.basket_hash = SimpleLazyObject(load_basket_hash)
 
         response = self.get_response(request)
         return self.process_response(request, response)
 
-    def process_response(self, request, response):
-        # Delete any surplus cookies
-        cookies_to_delete = getattr(request, 'cookies_to_delete', [])
-        for cookie_key in cookies_to_delete:
-            response.delete_cookie(cookie_key)
-
-        if not hasattr(request, 'basket'):
-            return response
+    @staticmethod
+    def is_basket_enabled(request):
```

Comment:
Is `staticmethod` really necessary? We call this method as generic class method and pass instance within this middleware class (`self.is_basket_enabled(request)`). E.g. can't find static method calls.